src/sqlancer/feldera/FelderaSchema.java

```diff
@@ -129,29 +114,144 @@ public FelderaExpression getRandomConstant(FelderaGlobalState globalState) {
 
             return FelderaConstant.getRandomConstant(globalState, this);
         }
+
+        public boolean isNumeric() {
+            return this.getPrimitiveType().isNumeric();
+        }
+
+        public FelderaCompositeDataType getElementType() {
+            return this.elementType;
+        }
+
+        public static FelderaCompositeDataType getBooleanType() {
+            return FelderaCompositeDataType.getRandomFromPrimitiveType(FelderaDataType.BOOLEAN);
+        }
+
+        public static FelderaCompositeDataType getRandomVarcharType() {
+            return FelderaCompositeDataType.getRandomFromPrimitiveType(FelderaDataType.VARCHAR);
+        }
+
+        public static FelderaCompositeDataType getRandomNumericType() {
+            FelderaDataType type = FelderaDataType.getRandomNumericType();
+            return FelderaCompositeDataType.getRandomFromPrimitiveType(type);
+        }
+
+        public FelderaDataType getPrimitiveType() {
+            return dataType;
+        }
+
+        public int getSize() {
+            return size;
+        }
+
+        public int getScale() {
+            return scale;
+        }
+
+        public boolean isArray() {
+            return dataType == FelderaDataType.ARRAY;
+        }
+
+        public static FelderaCompositeDataType getRandomFromPrimitiveType(FelderaDataType type) {
+            int size = -1;
+            int scale = -1;
+            switch (type) {
+            case FLOAT:
+                size = Randomly.fromOptions(32, 64);
+                return new FelderaCompositeDataType(type, size, scale);
+            case INT:
+                size = Randomly.fromOptions(8, 16, 32, 64);
+                return new FelderaCompositeDataType(type, size, scale);
+            case ARRAY:
+                return new FelderaCompositeDataType(type, FelderaCompositeDataType.getRandomWithoutNull());
+            case DECIMAL:
+                scale = (int) Randomly.getNotCachedInteger(0, 10);
+                size = (int) Randomly.getNotCachedInteger(scale, 25);
+                return new FelderaCompositeDataType(type, size, scale);
+            case CHAR:
+                size = (int) Randomly.getNotCachedInteger(1, 10);
+                return new FelderaCompositeDataType(type, size, scale);
+            case VARCHAR:
+                if (Randomly.getBoolean()) {
+                    size = (int) Randomly.getNotCachedInteger(1, 30);
+                }
+                return new FelderaCompositeDataType(type, size, scale);
+            default:
+                return new FelderaCompositeDataType(type, size, scale);
+            }
+        }
+
+        public static FelderaCompositeDataType getRandomWithoutNull() {
+            FelderaDataType type = FelderaDataType.getRandomNonNullType();
+            return FelderaCompositeDataType.getRandomFromPrimitiveType(type);
+        }
+
+        @Override
+        public String toString() {
```

Comment:
this suggests that Type should be an interface and these all subclasses.
What are the constraints imposed from SqlLancer?
Maybe you really should reuse the compiler type hierarchy.